### PR TITLE
Generate json for system methods

### PR DESCRIPTION
**Pull request by AtomicBlom**
Check out the systemMethods section of the gist for an example.
https://gist.github.com/AtomicBlom/f72597805b33e8b6ba7742d9aaf3ad40

### CODE REVIEW

File: src/extractMethods.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        if (isHeading()) return Number.parseInt(node.get(0).tagName[1]);
+        else throw "not a heading";
+    }
+    function isHeadingText(text: string) {
+        return isHeading() && getHeadingLevel() > targetHeadingLevel && node.text() === text;
+    }
+    function next() {
+        node = node.next();
+        while (isHeading() && node.is(":empty")) { //Methods seem to often have empty headers before the parameters and return types.
+            next();
```
Comment (abc-55) [inline src/extractMethods.ts]: This recursion is confusing. When `next()` returns, the condition is already false, so the `while` executes only once. The actual loop is created by recursion, not by `while`.

Change to `node = node.next();`

File: src/index.ts [hunk 2]
```diff
@@ -114,13 +139,30 @@ export namespace MinecraftScriptDocumentation {
...
             version: getVersion($),
+            systemMethods: [],
             components: [],
             events: {
                 client: { listening: [], triggerable: [] },
                 server: { listening: [], triggerable: [] }
             }
         };
 
+        var systemMethodCategories: {[key: string]: string} = {
```
Comment (abc-55) [inline src/index.ts]: Should be `const` (maybe i can configure tslint to check for things like that)

File: src/index.ts [hunk 3]
```diff
@@ -249,6 +291,52 @@ function extractType(table: Table): Type {
...
 
+function extractReturnType(table: Table): ReturnedType[] {
+    return table.map(row => {
+            const returnType: ReturnedType = {
+                type: row.cells[0],
+                value: row.cells[1],
+                description: row.cells[row.cells.length - 1] // default value may be missing
+            }
+
+            const arrayTypeCheck = (returnType.type || "").toString().toLocaleLowerCase();
```
Comment (abc-55) [inline src/index.ts]: This should be just `toLowerCase`. `toLocaleLowerCase` might for example convert uppercase i to dotless i in Turkish locale.

File: src/overrides.ts [hunk 4]
```diff
@@ -75,36 +119,61 @@ interface IScriptingDocumentationOverrides {
 }
 
 const overrideList: IScriptingDocumentationOverrides = cloneDeep(untypedOverrideList);
-const overridesVersion = overrideList.version;
 
 export function applyOverridesToDocumentation(documentation: MinecraftScriptDocumentation) {
```
Comment (abc-55) [inline src/overrides.ts]: This needs a better name to distinguish it from the `applyOverrides` function below. Maybe `applyDefaultOverrides`?

File: src/overrides.ts [hunk 5]
```diff
@@ -75,36 +119,61 @@ interface IScriptingDocumentationOverrides {
...
+        documentationVersion.revision !== overrideDefinitionVersion.revision ||
+        documentationVersion.build !== overrideDefinitionVersion.build) {
         console.error("-----------------------------------------------------------------------------");
         console.error("             The documentation version has been bumped!                      ");
         console.error("You will need to review the overrides to make sure they are still appropriate");
         console.error("          and then bump the version in type-overrides.json                   ");
         console.error("-----------------------------------------------------------------------------");
         console.info(`expected: ${JSON.stringify(overrideList.version)}`);
-        console.info(`actual  : ${JSON.stringify(documentationVersion)}`);
+        console.info(`actual  : ${JSON.stringify(overrideDefinitionVersion)}`);
```
Comment (abc-55) [inline src/overrides.ts:195]: expected should be `overrideDefinitionVersion`, actual should be `documentationVersion`

**Review by abc-55 — CHANGES_REQUESTED** [review]

File: src/extractMethods.ts (hunk 1 above)
Comment (AtomicBlom) [reply src/extractMethods.ts]: Good catch. That certainly wasn't my intention.

File: src/index.ts (hunk 2 above)
Comment (AtomicBlom) [reply src/index.ts]: Fixed.
We don't actually have tslint in the project, but I agree we really should, so I tried adding it but the 325 errors I got made me think that's a bit of a big change for this PR.
There's also a HEAP of warnings that I debate, such as enforced trailing commas, spaces between `<type> variable` when casting, and a 120 character line limit (I normally work to 160 nowadays)

File: src/index.ts (hunk 3 above)
Comment (AtomicBlom) [reply src/index.ts]: ooh, I'm glad you picked up on that. I'm pretty bad with locale issues. Fixed.

**Comment by AtomicBlom** [comment]
Hmm, looks like reviews on GitHub don't quite work the same as on Azure DevOps and it hides any comments that are left, so I'll resummarize here.

Good set of advice there. I've adopted all the ones mentioned.

We don't actually have tslint in the project, but I agree we really should, so I tried adding it but the 325 errors I got made me think that's a bit of a big change for this PR.
There's also a HEAP of warnings that I debate, such as enforced trailing commas, spaces between <type> variable when casting, and a 120 character line limit (I normally work to 160 nowadays)

**Pull request merged**